### DISCUSSION TITLE
Retieving data from inverter

### DISCUSSION THREAD

**Discussion by 2022Anto** [basking-in-the-sun2000/solar-logger#37] [General]
Thanks a lot for your project, it's really interesting!
What I'm going to ask are some clarification about the data connection: are you reteiving data directly from standard inverter's wlan on port 502 or did you install a rs485modbus to TCPmodbus? If yes did you use the Huawei Smart logger or a standard commercial product.
Any advice related to "_how to interface solarlogger with inverter_" is very welcome!
Bye
Anto

**Comment by bdkacz** [comment]
When You connect directly with wifi of inverter (SSID SUN2000....) you get ModbusTCP on port 502

**Reply by 2022Anto** [reply]
thanks!

**Comment by andysu** [comment]
afaik
Huawei Smart Logger has different register scheme.
SUN2000 has 2 different interface definition (v2 - SUN2000-*-KTL* and v3 SUN2000-*-M|L*)
v3 inverters has built-in wifi and modbusTCP - wifi and RTU -cable
v2 has built-in modbusRTU-cable

**Reply by 2022Anto** [reply]
thanks!

**Reply by basking-in-the-sun2000** [reply]
I had originally thought of using Huawei's Smartlogger, however, it wasn't compatible with my inverter. Thus I looked for a local logger and finally decided on coding one. I am accessing the inverter through my local network, rather than through the builtin AP

Think the registers for both maps are rather similar. Though you are right, I did use V3 for the logger. Have the idea my inverter is V1 (SUN2000--KTL-USL0) lol. Huawei hasn't updated the firmware in ages but still works with V3. Also, V3 should make it more compatible with current inverters.

However, the commercial units are different and tried to code it in _HuaweiEnt.default.py_. You can change which version of the definitions to use by setting **model** in _config.py._ Was asked to change the code to use multiple inverters, however, would be rather doubtful of how it works, since I couldn't test that.

The logger was supposed to be capable of pulling data from any inverter if you had the register map. In theory, by creating the file you should be able to poll the inverter, though might not be as nice as that. Probably the code is too biased toward the Huawei inverters

Since I don't use RTU, haven't tried it, but pymodbus should be capable. _modbustcp.py_ needs to be duplicated for the RTU calls and change the places that use _modbustcp.py_ in the code

**Reply by 2022Anto** [reply]
I tired to see the open ports on my inverter
**sudo nmap -n -PE -p- 192.168.200.2**
Starting Nmap 7.80 ( https://nmap.org ) at 2022-04-29 23:02 CEST
Nmap scan report for 192.168.200.2
Host is up (0.0087s latency).
All 65535 scanned ports on 192.168.200.2 are closed
MAC Address: xx:xx:xx:xx:xx:xx (xxxxxx)
Nmap done: 1 IP address (1 host up) scanned in 389.57 seconds

but no ports are open... that's the reason I'm not able to connect with my inverter

Do you believe it's feasible to install a converter like 
https://www.pusr.com/products/modbus-serial-to-ethernet-converters-usr-tcp232-410s.html
so I can use the data from the RS485 instead of inverter's WLAN?
Are you aware of any settings in Huawei inverter to enable port 502?
I'm using a
SOLAR2000-6KTL-L1
Fw: V200R001C00

thanks in advance for your support

**Comment by basking-in-the-sun2000** [comment]
Seems Huawei in recent firmwares started blocking tcp by default. Could be for improved security?

If you don't have port 502 open try this
https://forum.huawei.com/enterprise/en/modbus-tcp-guide/thread/789585-100027

**Reply by 2022Anto** [reply]
thanks for your advice, I followed the instructions but the port 502 is not open either on the default WLAN and on the dongle LAN

raspberrypi:~/solar-logger-master $ sudo nmap -n -PE -p 502 192.168.10.5
Starting Nmap 7.80 ( https://nmap.org ) at 2022-05-01 22:36 CEST
Nmap scan report for 192.168.10.5
Host is up (0.00033s latency).

PORT    STATE  SERVICE
502/tcp closed mbap
MAC Address: xx:xx:xx:xx (Huawei Technologies)

Nmap done: 1 IP address (1 host up) scanned in 0.65 seconds

Do you believe it's feasible to install a converter like
https://www.pusr.com/products/modbus-serial-to-ethernet-converters-usr-tcp232-410s.html
so I can use the data from the RS485?

**Reply by basking-in-the-sun2000** [reply]
Haven't tried it, but you might be able to use the rpi to connect to the rs485 bus. There are several hats that might work. My guess is that the code doesn't need a lot of changing. 

could this help? Seems huwaei opened another port 6607

https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-the-inverter

**Reply by 2022Anto** [reply]
Hello,
I contacted Huawei support and they provide me a different FW for Dongle: after upgrade I can connect to por 502

**sudo nmap -n -PE -p 502 -Pn 192.168.10.5**
Starting Nmap 7.80 ( https://nmap.org ) at 2022-05-06 13:38 CEST
Nmap scan report for 192.168.10.5
Host is up (0.00031s latency).

PORT    STATE SERVICE
**502/tcp open  mbap**
MAC Address: 28:68:D2:24:D8:CB (Huawei Technologies)

Nmap done: 1 IP address (1 host up) scanned in 0.76 seconds

I'll let you know about the solarlogger installation and usage!
Thanks a lot for your help!

**Reply by basking-in-the-sun2000** [reply]
😄

**Comment by 2022Anto** [comment]
Hello,
I tryed to run the main.py after checking that the connection is fine but I'm not able to see any data in response: I always receive 

```
anto@raspberrypi:~/solar-logger-master $ python3 main.py 
mailing not setup
starting
Thu May 26 23:57:54 2022
Loaded: Huawei
statring setup
opening modbus
192.168.10.5
opening db
setting db
setting params
main 1 error: list index out of range
main 2 error: list index out of range
SELECT min("Insulation") as "Insulation" FROM "Huawei_info" WHERE time >= 1653429600000000000 and time <= 1653516000000000000 and Insulation > 0
starting
in forecast
Thu May 26 23:57:56 2022
8
20
Fri May 27 00:22:56 2022
next forecast Thu May 26 07:45:00 2022
forecast error: 'forecasts'
solcast off-line
Fri May 27 00:27:57 2022
8
20
Fri May 27 00:52:57 2022
next forecast Thu May 26 07:45:00 2022
loop
mdobus read_registers error: Modbus Error: [Connection] ModbusTcpClient(192.168.10.5:502): Connection unexpectedly closed 0.000114 seconds into read of 8 bytes without response from unit before it closed connection
do_map error: 'int' object has no attribute 'isError'
trying to recover Model 1
mdobus read_registers error: Modbus Error: [Connection] ModbusTcpClient(192.168.10.5:502): Connection unexpectedly closed 1.762673 seconds into read of 8 bytes without response from unit before it closed connection
do_map error: 'int' object has no attribute 'isError'
trying to recover Model 2
read register error Model 3
read register error Model 4
Thu May 26 23:58:20 2022 Model
map looping
mdobus read_registers error: Modbus Error: [Connection] ModbusTcpClient(192.168.10.5:502): Connection unexpectedly closed 2.264459 seconds into unbounded read without response from unit before it closed connection
do_map error: 'int' object has no attribute 'isError'
trying to recover Model 1
mdobus read_registers error: Modbus Error: [Connection] ModbusTcpClient(192.168.10.5:502): Connection unexpectedly closed 2.265290 seconds into unbounded read without response from unit before it closed connection

```

Have you any suggestion?
I also tryed to retreive data with external clients but I'm never able to see any output
Thanks in advance!

**Reply by basking-in-the-sun2000** [reply]
moved to issue https://github.com/basking-in-the-sun2000/solar-logger/issues/42
